### PR TITLE
Output partial results

### PR DESCRIPTION
**Pull request by benv666**
This PR keeps track of the longest matching prefix found so far and adds some output to make it useful.
The purpose of this is so I can set a way too long prefix and have my machine crunch away at it until I actually need it and have the most beautiful prefix without worrying if it'll ever find it in time.
E.g.
```
./mine 0xa77E57EDF00d1234
Searching for 0xa77E57EDF00d1234 using 8 threads
Partial prefix match of len 1: Address: 0xad429891ff2121d47a3dfb3fec54efedf412a754 - salt: 0x04
Partial prefix match of len 2: Address: 0xa742c9391d131096e984d26620b4d8eea3dd37a9 - salt: 0x0307
Partial prefix match of len 3: Address: 0xa772f85bd0399b0d8627494c8a68a91c62d32d4d - salt: 0x2ff3
Partial prefix match of len 4: Address: 0xa77e39c1ebeac0fbd14acbf260f10510a3bab466 - salt: 0xe2ff
Partial prefix match of len 5: Address: 0xa77e54711ef71718cb73e3013c0520be4cb90362 - salt: 0x08401b
Partial prefix match of len 6: Address: 0xa77e577f7a0f48e1058a629cdde60f077b968139 - salt: 0xaf11ec
At salt 0x03d4ccc0... 5.00s (12.86M salts/sec), longest prefix found matches 6 chars using salt 0xaf11ec giving address: 0xa77e577f7a0f48e1058a629cdde60f077b968139
Partial prefix match of len 7: Address: 0xa77e57ef6eaa03518ee1ea68ec887e3f8d25e441 - salt: 0x04ecdd65
At salt 0x07ade080... 10.00s (12.91M salts/sec), longest prefix found matches 7 chars using salt 0x04ecdd65 giving address: 0xa77e57ef6eaa03518ee1ea68ec887e3f8d25e441
At salt 0x0bf1fc80... 15.00s (14.31M salts/sec), longest prefix found matches 7 chars using salt 0x04ecdd65 giving address: 0xa77e57ef6eaa03518ee1ea68ec887e3f8d25e441
At salt 0x0fd94740... 20.00s (13.10M salts/sec), longest prefix found matches 7 chars using salt 0x04ecdd65 giving address: 0xa77e57ef6eaa03518ee1ea68ec887e3f8d25e441
At salt 0x13f39980... 25.00s (13.77M salts/sec), longest prefix found matches 7 chars using salt 0x04ecdd65 giving address: 0xa77e57ef6eaa03518ee1ea68ec887e3f8d25e441
```
Now I can abort it and use the last found prefix instead of having to restart/continue and hoping it'll win the lottery in time.

NOTE: Performance wise it didn't seem to impact the ~13M-ish salts/sec on my machine, but perhaps this can be tuned :)

### CODE REVIEW

**Comment by jshufro** [comment]
Can we put this behavior behind a flag? by the way you can pass a starting salt to 'resume'

**Comment by benv666** [comment]
Flag is certainly possible, I'll see if I can add that later today. (assuming you mean a runtime flag)

Also the resume option is known, my point with restarting is that if you now find out your salt takes too long to compute, you'll need to restart with a smaller salt without my patch. With my patch you can just pick the longest found prefix without having to restart.

**Comment by benv666** [comment]
@jshufro Thoughts? :⁠-⁠)

File: main.c
Comment (jshufro) [inline main.c]: thread is accessing a global here- not safe

File: main.c
Comment (jshufro) [inline main.c]: cross-thread write contention here

File: main.c
Comment (jshufro) [inline main.c]: extremely unsafe memcpy

**Review by jshufro — CHANGES_REQUESTED** [review]
Alright, on deeper review, this is intolerably thread-unsafe. I can't merge it as-is, however, I like the idea.

That said, I don't like how much complexity it adds to the already-spaghetti-ish code that I slapped out in a weekend.

Stand by while I implement "plugin" support... thread safety is something I can't bake into it, though, since plugins will want to observe every mined address, and they exist in an arena ephemerally during the thread's main loop... so be prepared to learn the agony of systems engineering if you want to follow up on this

**Comment by jshufro** [comment]
Plugin support was added https://github.com/jshufro/rocketpool-vfvm/commit/7559a02904024bb7e2f83f35ba7dba50dfa17099

**Comment by benv666** [comment]
I'm giving it a shot rewriting it with your new plugin system while trying to keep thread-safety in mind, but while tinkering I noticed the new `on_iteration` callback does not provide the plugin with the corresponding salt.

The intermediate output does need it, since that's the entire point of this plugin.
Can we add salt as third component to the `on_iteration` callback?

**Comment by jshufro** [comment]
yeah feel free to add whatever fields you need

**Comment by benv666** [comment]
I made a stab, pulled it through `-fsanitize=thread` and `-fsanitize=leak` and that looked reasonable to me, but constructive criticism is very welcome :)

Another thing that crossed my mind would be to install an interrupt handler and print the partial results on CTRL-C, any thoughts on that?

**Comment by benv666** [comment]
@jshufro Once again I ask for your input :⁠-⁠)

File: plugins/partial-output/partial-output.h
Comment (jshufro) [inline plugins/partial-output/partial-output.h:1]: Remove this file

File: plugins/partial-output/partial-output.c
Comment (jshufro) [inline plugins/partial-output/partial-output.c]: this needs to be moved to main.c and call a new plugin->on_interrupt

File: plugins/partial-output/partial-output.c
Comment (jshufro) [inline plugins/partial-output/partial-output.c]: what's the performance like with these locks? there's probably a lock-free solution

File: plugins/partial-output/partial-output.c
Comment (jshufro) [inline plugins/partial-output/partial-output.c]:
```suggestion
	if (longest_prefix_len == 0) {
		return;
	}

	pthread_mutex_lock(&match_mutex);
	PRINTF("Longest prefix found matches %ld chars using salt ", longest_prefix_len);
	print_salt(longest_prefix_salt);
	printf(" giving address: 0x");
	for (size_t i = 0; i < ADDRESS_BYTES; i++)
		printf("%02x", longest_prefix_addr[i]);
	printf("\n");
	pthread_mutex_unlock(&match_mutex);
```

File: plugins/partial-output/partial-output.h
Comment (benv666) [reply plugins/partial-output/partial-output.h:1]: done

File: plugins/partial-output/partial-output.c
Comment (benv666) [reply plugins/partial-output/partial-output.c]: Since `my_prefix_len` only increases when a new longest match is found, this is imho not an issue. Only the first few matches will happen within seconds of starting the program, after that it'll be minutes/hours/days between this lock is used.

TDLR: negligible impact.

**Comment by benv666** [comment]
@jshufro Thanks again for the constructive criticism.

I've (hopefully) addressed your concerns, and possible introduced some new ones with the new `on_interrupt` construct I've created.

Note that I've opted to raise() the old signal in the signal handler, after tinkering a bit with setting the `done` bool to true in the sigint handler and testing with locks and their performance (including a construct which raised a `interrupted` sig_atomic_t causing the done to be toggled after a progress report, causing a 5-10 sec delay before quit). I feel like this has the best user experience, but feedback is again welcome.

File: plugins/partial-output/partial-output.c
Comment (jshufro) [inline plugins/partial-output/partial-output.c]:
```suggestion
	if (my_prefix_len <= longest_prefix_len) {
		return;
	}
	pthread_mutex_lock(&match_mutex);

	longest_prefix_salt = salt;
	memcpy(longest_prefix_addr, addr, ADDRESS_BYTES);
	atomic_store(&longest_prefix_len, my_prefix_len);

	pthread_mutex_unlock(&match_mutex);
```

not trying to be difficult, but early returns make for nicer c

**Review by jshufro — CHANGES_REQUESTED** [review]
I'm impressed you stuck with it all the way to this point. It's in really good shape now, just one nitpick.

Also, if you don't mind, before we merge, can you rewrite history into 2 commits?

1 which updates main.c, plugin.h and plugins/examples, and a second commit which adds partial-output?

File: plugins/partial-output/partial-output.c
Comment (benv666) [reply plugins/partial-output/partial-output.c]: Agreed, fixed

**Comment by benv666** [comment]
@jshufro Other than your requests no new changes, as far as I'm concerned this can be merged.
Thanks for the feedback and suggestions!

**Pull request merged**
